relayer/checkpoint.go

```diff
@@ -0,0 +1,206 @@
+// Copyright (C) 2024, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+package relayer
+
+import (
+	"container/heap"
+	"strconv"
+	"sync"
+
+	"github.com/ava-labs/avalanchego/utils/logging"
+	"github.com/ava-labs/awm-relayer/database"
+	"github.com/ava-labs/awm-relayer/utils"
+	"go.uber.org/zap"
+)
+
+//
+// checkpointManager commits keys to be written to the database in a thread safe manner.
+//
+
+type checkpointManager struct {
+	logger                   logging.Logger
+	database                 database.RelayerDatabase
+	writeSignal              chan struct{}
+	relayerID                database.RelayerID
+	queuedHeightsAndMessages map[uint64]*messageCounter
+	committedHeight          uint64
+	lock                     *sync.RWMutex
+	pendingCommits           *utils.UInt64Heap
+	finished                 chan uint64
+}
+
+func newCheckpointManager(logger logging.Logger, database database.RelayerDatabase, writeSignal chan struct{}, relayerID database.RelayerID) *checkpointManager {
+	h := &utils.UInt64Heap{}
+	heap.Init(h)
+	return &checkpointManager{
+		logger:                   logger,
+		database:                 database,
+		writeSignal:              writeSignal,
+		relayerID:                relayerID,
+		queuedHeightsAndMessages: make(map[uint64]*messageCounter),
+		lock:                     &sync.RWMutex{},
+		pendingCommits:           h,
+		finished:                 make(chan uint64),
+	}
+}
+
+func (cm *checkpointManager) run() {
+	go cm.handleFinishedRelays()
+	go cm.writeToDatabase()
+}
+
+func (cm *checkpointManager) writeToDatabase() {
+	for range cm.writeSignal {
+		cm.lock.RLock()
```

Comment:
I would encapsulate the body of this for loop into a helper function. That way, you can call lock and then defer the unlock. I think almost any critical section should always be it's own function

Reply:
Good call, I've updated the methods so that locks/unlocks are performed at function scope.